contracts/yellow/YellowClearingBase.sol

```diff
@@ -0,0 +1,207 @@
+        external
+        onlyRole(REGISTRY_MAINTAINER_ROLE)
+    {
+        require(address(_nextImplementation) == address(0), 'Next implementation already set');
+        require(
+            address(nextImplementation) != address(0) && address(nextImplementation) != _self,
+            'Invalid nextImplementation supplied'
+        );
+
+        require(
+            nextImplementation.hasRole(PREVIOUS_IMPLEMENTATION_ROLE, address(this)),
+            'Previous implementation role is absent'
+        );
+
+        _nextImplementation = nextImplementation;
+
+        emit NextImplementationSet(nextImplementation);
+    }
+
+    // Has participant
+    function hasParticipant(address participant) public view returns (bool) {
+        return _participantData[participant].status != ParticipantStatus.None;
+    }
+
+    // Get participant data
+    function getParticipantData(address participant)
+        external
+        view
+        returns (ParticipantData memory)
+    {
+        require(hasParticipant(participant), 'Participant does not exist');
+
+        return _participantData[participant];
+    }
+
+    // Recursively check that participant is not present in the registry
+    function requireParticipantNotRegistered(address participant) public view {
+        if (address(_nextImplementation) != address(0)) {
+            _nextImplementation.requireParticipantNotRegistered(participant);
+        }
+
+        require(!hasParticipant(participant), 'Participant already registered');
+    }
+
+    // Register participant with Pending status using signature by its Broker
+    function registerParticipant(address participant, ParticipantRegisterData calldata rData)
+        external
+    {
+        requireParticipantNotRegistered(participant);
+
+        require(
+            keccak256(abi.encode(participant)).toEthSignedMessageHash().recover(
+                rData.vaultBrokerSignature
+            ) == rData.vault.getBrokerAddress(),
+            'Signer is not participant vault broker'
+        );
+
+        _participantData[participant] = ParticipantData(
+            ParticipantStatus.Pending,
+            rData.vault,
+            block.number
+        );
+
+        emit ParticipantRegistered(participant);
+    }
+
+    // Set participant data
+    function setParticipantData(address participant, ParticipantData memory data)
+        external
+        onlyRole(REGISTRY_MAINTAINER_ROLE)
+    {
+        require(participant != address(0), 'Invalid participant address');
+
+        require(
+            _participantData[participant].status != ParticipantStatus.Migrated,
+            'Participant migrated'
+        );
+
+        _participantData[participant] = data;
+
+        emit ParticipantDataSet(participant, data);
+    }
+
+    // Migrate participant
+    function migrateParticipant() external {
+        require(address(_nextImplementation) != address(0), 'Next version is not set');
+
+        address participant = msg.sender;
```

Comment:
Instead of using `msg.sender` for participant, take it as parameter, with a signed payload by the participant address.

This allow participant address to remain "virtual" (without any funds, acting only as in ID in the network).
If we use msg.sender here, this force participant address to have some chain tokens to pay for tx gas, which would be anoying to always topup this address to perform migration.
Instead we let someone else pay for it, and proove he actual own the address by providing a signed payload by this participant address.

Reply:
9d5f8d9d6e617f94de7b520b003e39dbaae5f33a